src/algorithms/boundary.rs

```diff
@@ -0,0 +1,515 @@
+//! Algorithms for handling port boundaries in a graph.
+
+use std::collections::{HashMap, HashSet};
+
+use super::toposort::toposort;
+use crate::{Direction, LinkView, NodeIndex, PortIndex, PortView};
+
+/// A port boundary in a graph.
+///
+/// Defined
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Boundary {
+    /// The ordered list of incoming ports in the boundary.
+    inputs: Vec<PortIndex>,
+    /// The ordered list of outgoing ports in the boundary.
+    outputs: Vec<PortIndex>,
+}
```

Comment:
Would probably make sense to also use this `struct` for `Subgraph<G>` in [src/view/subgraph.rs](https://github.com/CQCL/portgraph/blob/0722b985a31de5eca0906dfb590cab82a4ec55da/src/view/subgraph.rs#L57)? If so, you might consider moving this definition out of `src/algorithms/`.

This might be a separate PR, but maybe it's an indication that `Boundary` shouldn't live within the `algorithms` module?